plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -698,89 +874,10 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
-      case vote_result_t::duplicate:
-         break;
-      case vote_result_t::unknown_public_key:
-      case vote_result_t::invalid_signature:
-      case vote_result_t::max_exceeded:
-      case vote_result_t::unknown_block:
-         return;
-      default:
-         assert(false); // should never happen
-      }
-
-      if (!active_auth && !pending_auth) {
-         elog("vote signal contains no valid authority ${m}", ("m", msg));
-         return;
-      }
-
-      auto now = fc::time_point::now();
-      account_name auth_desc = active_auth ? to_account_name_safe(active_auth->description)
-                                           : to_account_name_safe(pending_auth->description);
-      if (auth_desc.empty()) {
-         ilog("Finalizer authority description is not a valid producer name ${d}",
-              ("d", active_auth ? active_auth->description : pending_auth->description));
-         // running with core contract that does not associate finalizer_authority->description with a producer
-         // reset times otherwise the producer would pause
-         _last_other_vote_received.store(now, std::memory_order_relaxed);
-         _last_producer_vote_received.store(now, std::memory_order_relaxed);
-         return;
-      }
-
-      if (_producers.contains(auth_desc)) { // _producers not modified, thread safe
-         _last_producer_vote_received.store(now, std::memory_order_relaxed);
-      } else {
-         _last_other_vote_received.store(now, std::memory_order_relaxed);
-      }
-   }
-
-   enum class implicit_pause {
-      not_paused,   // not implicitly paused
-      prod_paused,  // paused due to not receiving vote associated with producer
-      other_paused  // paused due to not receiving any vote from finalizers not associated with producer
-   };
-   implicit_pause implicitly_paused() const {
-      if (_producers.empty() || _production_pause_vote_timeout.count() == 0)
-         return implicit_pause::not_paused;
-      if (!_is_savanna_active)
-         return implicit_pause::not_paused; // no implicit pause in legacy
-      if (_producers.contains(eosio::chain::config::system_account_name)) // disable implicit pause for eosio
-         return implicit_pause::not_paused;
-
-      auto last_producer_vote_received = _last_producer_vote_received.load(std::memory_order_relaxed);
-      auto last_other_vote_received    = _last_other_vote_received.load(std::memory_order_relaxed);
-
-      // need a vote within timeout of last accepted block
-      if (_is_producer_active_finalizer &&
-          _accepted_block_time - last_producer_vote_received > _production_pause_vote_timeout) {
-         return implicit_pause::prod_paused;
-      }
-      if (_other_active_finalizers &&
-          _accepted_block_time - last_other_vote_received    > _production_pause_vote_timeout) {
-         return implicit_pause::other_paused;
+         if (_is_savanna_active) {
+            _implicit_pause_vote_tracker.activate();
```

Comment:
There would be no need for this in the simplified `production_pause_vote_tracker` class I am proposing. In fact, `activate` and `init` member functions (among others) would go away entirely.

Reply:
The tracking of active finalizers is rather expensive, seems like we should avoid tracking them on validation nodes. I've wrapped your `production_pause_vote_trakcer` in `implicit_production_pause_vote_tracker`.